### ISSUE TITLE
Build Failure Since Upgrade to 2.1.400

### ISSUE THREAD

**Issue by nguerrera** [microsoft/msbuild#3627]
_From @jamiebrynes7 on August 15, 2018 11:5_

## Steps to reproduce
1. Install dotnet 2.1.400
2. Clone https://github.com/spatialos/UnityGDK
3. Ensure that `global.json` has the version set to `2.1.400`
3. Run `./ci/codegen.sh` or run the following command directly from the root of the project:
```
dotnet run -p code_generator/GdkCodeGenerator/GdkCodeGenerator.csproj -- --schema-compiler-path=tools/schema_compiler/win/schema_compiler.exe --schema-path=schema --schema-path=build/dependencies/schema/standard_library --json-dir=workers/unity/Temp/ImprobableJson --native-output-dir=workers/unity/Assets/Generated/Source --network-types-output-dir=workers/unity/Assets/Improbable.Generated.NetworkTypes/Generated
```

## Expected  behavior
The build succeeds and the project is ran.

## Actual behavior
The build fails with the following message:
```
C:\Workspace\Develop\UnityGDK\code_generator\GdkCodeGenerator\GdkCodeGenerator.csproj(51,5): error MSB4018: The "TransformAllTask" task failed unexpectedly.
C:\Workspace\Develop\UnityGDK\code_generator\GdkCodeGenerator\GdkCodeGenerator.csproj(51,5): error MSB4018: System.IO.FileLoadException: Could not load file or assembly 'System.CodeDom, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)
C:\Workspace\Develop\UnityGDK\code_generator\GdkCodeGenerator\GdkCodeGenerator.csproj(51,5): error MSB4018: File name: 'System.CodeDom, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
C:\Workspace\Develop\UnityGDK\code_generator\GdkCodeGenerator\GdkCodeGenerator.csproj(51,5): error MSB4018:    at Mono.TextTemplating.TemplateGenerator..ctor()
C:\Workspace\Develop\UnityGDK\code_generator\GdkCodeGenerator\GdkCodeGenerator.csproj(51,5): error MSB4018:    at Improbable.TextTemplating.TransformAllTask.Execute()
C:\Workspace\Develop\UnityGDK\code_generator\GdkCodeGenerator\GdkCodeGenerator.csproj(51,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
C:\Workspace\Develop\UnityGDK\code_generator\GdkCodeGenerator\GdkCodeGenerator.csproj(51,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
C:\Workspace\Develop\UnityGDK\code_generator\GdkCodeGenerator\GdkCodeGenerator.csproj(51,5): error MSB4018:
C:\Workspace\Develop\UnityGDK\code_generator\GdkCodeGenerator\GdkCodeGenerator.csproj(51,5): error MSB4018:

The build failed. Please fix the build errors and run again.
```

Where `TransformAllTask` is a pre-build step to generate `cs` files from `tt` which are then used at execution time. `TransformAllTask` is in `Improbable.TextTemplating.dll` that is pinned, along with its dependencies in the `GdkCodeGenerator` project. (Rebuilding the dlls and replacing doesn't affect the result).

This build succeeds when ran with dotnet 2.1.30x

## Environment data
`dotnet --info` output:

```
λ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.400
 Commit:    8642e60a0f

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.400\

Host (useful for support):
  Version: 2.1.2
  Commit:  811c3ce6c0

.NET Core SDKs installed:
  2.1.103 [C:\Program Files\dotnet\sdk]
  2.1.105 [C:\Program Files\dotnet\sdk]
  2.1.301 [C:\Program Files\dotnet\sdk]
  2.1.400 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

_Copied from original issue: dotnet/cli#9839_

**Comment by nguerrera** [comment]
There's a 4.0.0.0 System.CodeDom.dll that comes in from MSBuild to the CLI starting in 15.8/2.1.400 and this seems to prevent a custom task from loading a newer version.

cc @jeffkl @rainersigwald @AndyGerlicher

**Comment by nguerrera** [comment]
https://github.com/Microsoft/msbuild/pull/3175/files#diff-4d3fc7780500131817fceb95ddcea3feR957

**Comment by rainersigwald** [comment]
I was hoping that modifying your project to reference the same version of `System.CodeDOM` that MSBuild does (package version `4.4.0`, assembly version `4.0.0.0`) would resolve the problem, but it does not because (checked-in) `Mono.TextTemplating.dll` refers to `System.CodeDom, Version=4.0.1.0`, which is causing the real load problem.

I think the long-term solution here is #1754, but that's not something that can be worked around easily.

@jamiebrynes7 I was able to get your repro working in a few steps:

1. Build Mono.TextTemplating (from https://github.com/mono/t4/commit/630491449f219b3f5e7674e054d92794dd091d74).
1. Copy the `netstandard2.0` output to `code_generator/GdkCodeGenerator/dependencies/Improbable.TextTemplating/Mono.TextTemplating.dll` in your repo.
1. Build

This works for .NET Core since Mono.TextTemplating recently added .NET Core support and depends on the same version of the package that we do there.

Unfortunately it fails with a loader error on full framework `MSBuild.exe`, because Mono.TextTemplating added public key token (https://github.com/mono/t4/commit/1c3ed573badfb76c339206acffbe97dbe156e220), so the .NET loader fails with:

```
LOG: Attempting download of new URL file:///s:/work/UnityGDK/code_generator/GdkCodeGenerator/dependencies/Improbable.TextTemplating/Mono.TextTemplating.DLL.
WRN: Comparing the assembly name resulted in the mismatch: PUBLIC KEY TOKEN
```

That could be resolved by either patching your local build of Mono.TextTemplating to drop the PKT, or rebuilding `Improbable.TextTemplating.dll` against the new assembly.

Or you could have two copies of the assemblies, one as-is and one for .NET Core that uses all the new stuff.

**Comment by jamiebrynes7** [comment]
@rainersigwald 

Thanks for the writeup - I've been able to get the project building and running using your workaround listed above.

Strangely - that version of Mono.TextTemplating was the exact version I pulled in as source when setting up the project (as the NuGet version is slightly outdated).

We only expect users to run the project through.NET Core anyway so that works perfectly! 

Thanks again! :smile:

**Comment by nguerrera** [comment]
I just hit this with the port of presentationbuildtasks to .net core

**Comment by nguerrera** [comment]
I have worked around it by using package version 4.4.0 for presentationbuildtasks, but I think we should prioritize #1754. Maybe we don't even need this bug and consider it a dupe of #1754.

**Comment by nguerrera** [comment]
#3991 updates to 4.5.0, I think we can call that good enough for this bug, but the real fix is #1754.

**Comment by nguerrera** [comment]
https://github.com/dotnet/toolset/pull/1127 will fix this for CodeDom and other fx assemblies such that the SDK will always lift codedom up to latest now, but we should still do #1754.
